specification/dartLangSpec.tex

```diff
@@ -10474,84 +10508,158 @@ \subsection{Assignment}
 \LMLabel{assignment}
 
 \LMHash{}%
-An assignment changes the value associated with a mutable variable or property.
+An assignment changes the value associated with a mutable variable,
+or invokes a setter.
 
 \begin{grammar}
 <assignmentOperator> ::= `='
   \alt <compoundAssignmentOperator>
+
+<compoundAssignmentOperator> ::= `*='
+  \alt `/='
+  \alt `~/='
+  \alt `\%='
+  \alt `+='
+  \alt `-='
+  \alt `\ltlt='
+  \alt `\gtgt='
+  \alt `\gtgtgt='
+  \alt `\&='
+  \alt `^='
+  \alt `|='
+  \alt `??='
 \end{grammar}
 
 \LMHash{}%
-\Case{\code{$v$ = $e$}}
-%% TODO(eernst): This _only_ works if we assume that `v = e` has already
-%% been expanded to `this.v = e` "when that's the right thing to do".
-%% Otherwise `denotes` below cannot be interpreted as the result of a lookup,
-%% and we have no precise alternative which would work. We might be able
-%% to repair this by giving a definition of `denotes` somewhere.
-Consider an assignment $a$ of the form \code{$v$ = $e$},
-where $v$ is an identifier or an identifier qualified by an import prefix,
-and $v$ denotes a variable (\ref{variables}) or \code{$v$=} denotes a setter
-(\commentary{which may be declared explicitly or induced by an instance variable, etc}).
-Let $T$ be the static type of $v$ when $v$ denotes a variable,
-otherwise let $T$ be the static type of the formal parameter of the setter \code{$v$=}.
-It is a compile-time error if the static type of $e$ may not be assigned to $T$.
-The static type of $a$ is the static type of $e$.
+\Case{\code{\id{} = $e$}}
+Consider an assignment $a$ of the form \code{\id{} = $e$},
+where \id{} is an identifier.
+
+If \id{} is a local variable $v$
+(\commentary{which may be a formal parameter}),
+a compile-time error occurs if $v$ is final
+or if the static type of $e$ is not assignable to the declared type of $v$.
+
+Otherwise
+(\commentary{when \id{} is not a local variable}),
+a lexical lookup of \code{\id=}
+(\ref{lexicalLookup})
+is performed from the location of \id.
+
+\begin{itemize}
+\item
+  When the lexical lookup yields a declaration $D$,
+  it is guaranteed to be a setter
+  (\commentary{that may be explicit or induced implicitly by a variable})
+  because other declarations do not have a name
+  of the form \code{\id=}.
+
+  If $D$ is the declaration of a static setter in class $C$
+  then $a$ is treated as
+  (\ref{notation})
+  the assignment \code{$C$.\id{} = $e$}.
+
+  \commentary{%
+    Further analysis as well as evaluation of \code{$C$.\id{} = $e$}
+    proceeds as specified elsewhere.%
+  }
+
+  Otherwise, a compile-time error occurs,
+  unless the static type of $e$ is assignable to the parameter type of $D$.
+\item
+  When the lexical lookup yields a member signature,
+  $a$ is treated as
+  (\ref{notation})
+  \code{\THIS.\id{} = $e$}.
+
+  \commentary{%
+    This occurs when no declaration named \code{\id=} is in scope,
```

Comment:
> This also occurs  ...

(because otherwise it sounds like it only occurs then, and not when you actually find an instance setter first).

Reply:
Adjusted to avoid saying whether the declaration was found, stating instead the important guarantees for this situation. Then it's up to the inquisitive reader to look at \ref{lexicalScope} to verify that it's always _as intended_ that we are now accessing an instance member rather than any other kind of entity.